mm2src/coins/tendermint/tendermint_coin.rs

```diff
@@ -4,76 +4,115 @@ use super::tendermint_native_rpc::*;
 use cosmrs::tx::{self, Fee, Msg, Raw, SignDoc, SignerInfo};
-use cosmrs::{AccountId, Any, Coin, Denom};
+use cosmrs::{AccountId, Any, Coin, Denom, ErrorReport};
+use crypto::privkey::key_pair_from_secret;
 use derive_more::Display;
 use futures::lock::Mutex as AsyncMutex;
 use futures::{FutureExt, TryFutureExt};
 use futures01::Future;
+use hex::FromHexError;
 use keys::KeyPair;
 use mm2_core::mm_ctx::MmArc;
 use mm2_err_handle::prelude::*;
 use mm2_number::MmNumber;
+use parking_lot::Mutex;
 use prost::{DecodeError, Message};
 use rpc::v1::types::Bytes as BytesJson;
 use serde_json::Value as Json;
+use std::collections::HashMap;
 use std::convert::TryFrom;
 use std::ops::Deref;
 use std::str::FromStr;
 use std::sync::Arc;
-
-const TIMEOUT_HEIGHT_DELTA: u64 = 100;
+use std::time::Duration;
+use uuid::Uuid;
+
+// ABCI Request Paths
+const ABCI_GET_LATEST_BLOCK_PATH: &str = "/cosmos.base.tendermint.v1beta1.Service/GetLatestBlock";
+const ABCI_GET_BLOCK_BY_HEIGHT_PATH: &str = "/cosmos.base.tendermint.v1beta1.Service/GetBlockByHeight";
+const ABCI_SIMULATE_TX_PATH: &str = "/cosmos.tx.v1beta1.Service/Simulate";
+const ABCI_QUERY_ACCOUNT_PATH: &str = "/cosmos.auth.v1beta1.Query/Account";
+const ABCI_QUERY_BALANCE_PATH: &str = "/cosmos.bank.v1beta1.Query/Balance";
+const ABCI_GET_TX_PATH: &str = "/cosmos.tx.v1beta1.Service/GetTx";
+const ABCI_QUERY_HTLC_PATH: &str = "/irismod.htlc.Query/HTLC";
+const ABCI_GET_TXS_EVENT_PATH: &str = "/cosmos.tx.v1beta1.Service/GetTxsEvent";
+
+/// 0.25 is good average gas price on atom and iris
+const DEFAULT_GAS_PRICE: f64 = 0.25;
+pub(super) const TIMEOUT_HEIGHT_DELTA: u64 = 100;
 pub const GAS_LIMIT_DEFAULT: u64 = 100_000;
-pub const TX_DEFAULT_MEMO: &str = "";
+pub(crate) const TX_DEFAULT_MEMO: &str = "";
+
+// https://github.com/irisnet/irismod/blob/5016c1be6fdbcffc319943f33713f4a057622f0a/modules/htlc/types/validation.go#L19-L22
+const MAX_TIME_LOCK: i64 = 34560;
+const MIN_TIME_LOCK: i64 = 50;
 
-#[derive(Clone, Debug, PartialEq, Serialize)]
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
 pub struct TendermintFeeDetails {
-    coin: String,
-    amount: BigDecimal,
-    gas_limit: u64,
+    pub coin: String,
+    pub amount: BigDecimal,
+    pub gas_limit: u64,
 }
 
 #[derive(Clone, Debug, Deserialize, Serialize)]
 pub struct TendermintProtocolInfo {
     decimals: u8,
     denom: String,
-    account_prefix: String,
+    pub account_prefix: String,
     chain_id: String,
+    gas_price: Option<f64>,
 }
 
-#[derive(Clone, Deserialize)]
-pub struct TendermintActivationParams {
-    rpc_urls: Vec<String>,
+#[derive(Clone)]
+pub struct ActivatedTokenInfo {
+    decimals: u8,
+    denom: Denom,
 }
 
 pub struct TendermintCoinImpl {
     ticker: String,
-    rpc_client: HttpClient,
+    /// TODO
+    /// Test Vec<String(rpc_urls)> instead of HttpClient and pick
+    /// better one in terms of performance & resource consumption on runtime.
+    rpc_clients: Vec<HttpClient>,
```

Comment:
I prefer adding special structures that incapsulates some logic like URL validation etc.
Isn't more readable code more important than several nanosecond we save on operations?
I like this `HttpClient`

Reply:
I don't know how different performance there will be or how different resource usage there  will be. I haven't checked how HttpClient gets created and how it stored in memory. That's why I left TODO there.